### DISCUSSION TITLE
How override selected row styles?

### DISCUSSION THREAD

**Discussion by iakovalgaev** [table-library/react-table-library#35] [Q&A] [answered]
I want to customize selected rows but I didn't find a way to do it. Even my `<Row className='some' />`  className prop is added before the library classes. And I can't override inline styles normally.

**Comment by rwieruch** [comment, marked as answer]
Hi @bezvrednyy Thanks for asking:

If you want to give a row a specific CSS class, do it this way:

```
<Row className="foo" ... >
```

Then if you want to style all Row components with this CSS class, specify it in the `THEME` that is passed to `const = useTheme(THEME)`:

```
const THEME = {
  Row: `
    &.foo {
      background-color: pink;
    }
  `
};
```

Regarding the selected state on a row, when you inspect the HTML elements in the browser, you will see that the Row component receives the following **optional*** CSS classes based on its select status:

```
row-select-selected
row-select-single-selected
```

While the former is applied for multi-select, the latter gets applied for single select. So if you want to style a row for multi-select, do it the following way:

```
const THEME = {
  Row: `
    &.row-select-selected {
      background-color: red;
    }
  `
};
```

For single select it would look like this:

```
const THEME = {
  Row: `
    &.row-select-single-selected {
      background-color: blue;
    }
  `
};
```

If you want to style both cases the same way, then use the selectors this way:

```
const THEME = {
  Row: `
    &.row-select-selected, &.row-select-single-selected {
      background-color: orange;
    }
  `
};
```

Let me know if this answer helped you!

**Reply by iakovalgaev** [reply]
@rwieruch, thank you for the prompt response. I love you! This shed light on the styling.

But ideally, I would like not to use useTheme(), but to have my className on `<Row>`, `<Cell/>`, `<Body/>`, redefined the built-in styles. For example. I would like to use tailwind for this: `<Row className='bg-blue-500'/>`. Also, it would greatly simplify the support of the dark theme.

And for selected, I want to have some external property that I could rely on. There is a similar thing in the headless ui:

```
function MyListbox() {
  const [selectedPerson, setSelectedPerson] = useState(people[0])

  return (
    <Listbox value={selectedPerson} onChange={setSelectedPerson}>
      <Listbox.Button>{selectedPerson.name}</Listbox.Button>
      <Listbox.Options>
        {people.map((person) => (
          /* Use the `active` state to conditionally style the active option. */
          /* Use the `selected` state to conditionally style the selected option. */
          <Listbox.Option key={person.id} value={person} as={Fragment}>
            {({ active, selected }) => (
              <li
                className={`${
                  active ? 'bg-blue-500 text-white' : 'bg-white text-black'
                }`}
              >
                {selected && <CheckIcon />}
                {person.name}
              </li>
            )}
          </Listbox.Option>
        ))}
      </Listbox.Options>
    </Listbox>
  )
}
```

It seems to me not very good to depend on class names, because if the library changes class names in the next version, styles will be broken for users, and developers will find out about it too late. And when changing props, parameters, TypeScript will immediately report this

These are my wishes and ideas, I hope they will be useful.
Thank you for the answer. I didn't find it in the documentation, it would be great to add it to the Features->Select section.

**Reply by iakovalgaev** [reply]
Another problem. Now it is impossible even in useTheme to redefine the white-space property for the HeaderCell. You must use `<style>` or !important.

**Reply by rwieruch** [reply]
> Another problem. Now it is impossible even in useTheme to redefine the white-space property for the HeaderCell. You must use <style> or !important.

I tried to use a minimal set of sensible defaults for the initial styling of the table. But you are right in this case (and perhaps there are other cases as well). Would you agree that it would be better to ship the components without any style and enable users to activate the status quo as opt-in baseline theme? Then one could start off with the bare bones components.

**Reply by rwieruch** [reply]
Regarding the select state. What you can do at the moment is using the select object to find out about which items are selected:

```
const select = useRowSelect(data);

console.log(select.state);

// ... then in the Row component when you assign the className conditionally:

<Row className={select.state.ids.includes(item.id) ? "multi-selected" : "not-selected"}  ... >

// ... or for single select

<Row className={select.state.id === item.id ? "single-selected" : "not-selected"}  ... >
```

What I also could think of is adding some more styling options to the hook itself:

```
const select = useRowSelect(data, { onChange }, {
  classNameSelected: 'my-custom-selected-class',
  classNameSingleSelected: 'my-custom-single-selected-class',
  classNameMultiSelected: 'my-custom-multi-selected-class',
});
```

Do you think this would be useful or do the previous recommendations already help?

**Reply by iakovalgaev** [reply]
The previous recommendations helped. Thanks.

There was another problem. If the cells of the last column are set to `display: none`. The `width` and `min-width` for specified in useTheme are redefined. 
By what principle does this happen and how can I fix it?
I want to add row actions with display: none and position: absolute. They were displayed when hovering over the structure and at the same time they were NOT a column and did not take up the width.

![image](https://user-images.githubusercontent.com/61377069/161712453-0317d636-d744-4125-bbaa-d31a8e27cd6a.png)

**Reply by rwieruch** [reply]
If you want to hide a cell, check out the built-in feature: https://react-table-library.com/?path=/docs/compact-table--column-hiding Hope this helps.

**Reply by iakovalgaev** [reply]
I want to do a hover effect.
![image](https://user-images.githubusercontent.com/61377069/161733956-ec467644-35ec-4ee1-a50f-2522050a84db.png)

Speaking of hiding columns. I don't like the built-in way. It works fast, but does not recalculate the table. If I have a table with a horizontal scroll, I want its width to decrease when hiding the column. But this doesn't happen, so I use external filtering by changing columns.

**Reply by rwieruch** [reply]
Does this story help? There you got a hover effect. Just added it with the latest deploy. https://react-table-library.com/?path=/docs/misc-actions--base

@bezvrednyy please open a new discussion if the topic changed in the future. Thanks for all your input :)
